### PR TITLE
Drupal 11 readiness.

### PR DESCRIPTION
**Pull request by apathak18**
Prepare drupal-graphql module ready for Drupal 11

### CODE REVIEW

**Comment by klausi** [comment]
Thanks, if we start to support Drupal 11 we should enable it in our testing. Can you enable it in testing.yml?

**Comment by apathak18** [comment]
> Thanks, if we start to support Drupal 11 we should enable it in our testing. Can you enable it in testing.yml?

@klausi Initial change in the workflow to include D10.3.x and D11.0.x 
Now Jobs are failing and will push changes as per the failed jobs.

File: composer.json
Comment (klausi) [inline composer.json]: should be 8.1 as Drupal 10 requires that.

File: composer.json
Comment (klausi) [inline composer.json]: do we need to bump the graphql-php dependency?

File: .github/workflows/testing.yml
Comment (klausi) [inline .github/workflows/testing.yml]: i think it is enough to test 10.2.x just on PHP 8.1, and only test 10.3 on all the versions

File: .github/workflows/testing.yml
Comment (klausi) [inline .github/workflows/testing.yml]: not sure we can do that - for now phpstan should run on the latest drupal prod release which is 10.3.

**Review by klausi — CHANGES_REQUESTED** [review]
Thanks!

The test fails Call to undefined method `GraphQL\Server\ServerConfig::setPersistentQueryLoader()` probably happen because you updated the graphql library.

If we are forced to do that then we need to fix our API calls as well.

**Comment by klausi** [comment]
I just merged #1410 to enable Drupal 10.3 testing. So in this PR we only need to add one extra run for Drupal 11 without PHPStan.

File: composer.json
Comment (apathak18) [reply composer.json]: I guess yes the last tag of 14.11.0 supports bare minimum package with php minimum support of 7.1 and phpunit maximum support is ^8.5
whereas latest tag for ^15.12.5 minimum php support is 7.4 and phpunit 9.5 and 10, also added few new packages under require-dev section.
I suggest we should bump this package version to make things work with D10 and above.

**Comment by Kingdutch** [comment]
Thanks for opening this! 

I do believe we should split this PR up into multiple PRs and possibly over multiple releases.

For example the webonyx/graphql-php dependency was flagged already as breaking change in https://github.com/drupal-graphql/graphql/issues/1335 so if that's a blocker for Drupal 11 then we likely need to release GraphQL 5.0.0 with Drupal 10.x and 11 compatibility.

We should also evaluate the other dependency upgrades and determine whether they're needed or nice to have. Ideally doing them in a way where extending modules can focus on one change at a time.

I think it's important for a smooth upgrade experience that we apply Symfony's upgrade philosophy here. We should ensure that consumers can always focus on upgrading one thing at a time. For example, they can focus on upgrading the PHP version and then separately the GraphQL package before making the jump to Drupal 11. Or Drupal 11 first and then the GraphQL package.

I know from experience that modules that force adopters to upgrade multiple composer packages at the same time and/or the PHP version as well, are a lot harder than having only one of those be included in a release.

With the insights that you already have of what might need to be upgraded it'd be great to open up an issue that outlines which changes we need to make so we can plan the right approach.

**Comment by apathak18** [comment]
@Kingdutch @klausi I've rebased the PR and revert the webnoyx package to v14 as v15 starts failing all tests.

Right now tests are passing for all D10.x.x versions except 11.0.x (Am I missing something or is there any configuration for database server which needs to be updated)
See:https://github.com/drupal-graphql/graphql/actions/runs/9996316696/job/27630424860?pr=1407#step:11:1257

**Comment by vishalkhode1** [comment]
> @Kingdutch @klausi I've rebased the PR and revert the webnoyx package to v14 as v15 starts failing all tests.
> 
> Right now tests are passing for all D10.x.x versions except 11.0.x (Am I missing something or is there any configuration for database server which needs to be updated) See:https://github.com/drupal-graphql/graphql/actions/runs/9996316696/job/27630424860?pr=1407#step:11:1257

@apathak18 The tests are failing because of old version of Ubuntu, you should update `.github/workflows/testing.yml` file and change `runs-on: ubuntu-latest` to `runs-on: ubuntu-24.04`

**Comment by apathak18** [comment]
@Kingdutch @klausi 11.0.x pipeline reports phpstan issues ~~ Mentioned below  are Changelog of those deprecations:

1. https://www.drupal.org/node/2999981 (format_size)
2. https://www.drupal.org/node/3363700 (file_validate)

Can we bump the drupal minimum version to D10.2? Otherwise we need to handle these deprecations to support backward compatibility.

**Review by vishalkhode1 — APPROVED** [review]

**Comment by apathak18** [comment]
Some attachments based on validating the changes on local.
Screenshot 2024-07-25 at 10 54 01 PM
Screenshot 2024-07-25 at 10 54 08 PM
Screenshot 2024-07-25 at 10 54 21 PM

File: src/Plugin/GraphQL/DataProducer/Entity/Fields/Image/ImageDerivative.php
Comment (vishalkhode1) [inline src/Plugin/GraphQL/DataProducer/Entity/Fields/Image/ImageDerivative.php]: @apathak18 We've updated condition from `empty` to `!isset`. The empty condition check, returns `TRUE` for empty string value and returns `FALSE` for `NULL` value, whereas `isset` return `TRUE` for `NULL` value and return `FALSE` for empty string value. If we want to limit condition only for `NULL` value, then better we should update it `$width == NULL` for better code visibility.

File: tests/src/Kernel/Framework/UploadFileServiceTest.php
Comment (vishalkhode1) [inline tests/src/Kernel/Framework/UploadFileServiceTest.php:21]: @apathak18 Why we removed it here ? It's better we add reason here, so maintainers are aware of on why this change was made.

File: src/GraphQL/Utility/FileUpload.php
Comment (vishalkhode1) [inline src/GraphQL/Utility/FileUpload.php]: @apathak18 We are returning response twice i.e first validating if path is valid and then next time we are validating image resolution. It would be great, if we can perform all validations and return all errors at once so that user is aware of what all errors they need to fix.

**Review by vishalkhode1 — CHANGES_REQUESTED** [review]

File: tests/src/Kernel/Framework/UploadFileServiceTest.php
Comment (apathak18) [reply tests/src/Kernel/Framework/UploadFileServiceTest.php:21]: Parent class graphqlTestBase already have file module.
And graphql_file_validate is only testing for hook_file_validate which is deprecated in D10.2 and we are using file.validator service as a replacement.

File: src/GraphQL/Utility/FileUpload.php
Comment (apathak18) [reply src/GraphQL/Utility/FileUpload.php]: Done, thanks!!

**Review by vishalkhode1 — APPROVED** [review]

File: .github/workflows/testing.yml
Comment (klausi) [inline .github/workflows/testing.yml:11]: any reason why we need to change this? Can we leave it at ubuntu-latest?

File: .github/workflows/testing.yml
Comment (klausi) [inline .github/workflows/testing.yml:27]: we should also test core 10.3 with PHP 8.3. Can you add php 8.3 to the php versions above?

File: src/GraphQL/Utility/FileUpload.php
Comment (klausi) [inline src/GraphQL/Utility/FileUpload.php]: any reason why this comment was changed? The previous description was better and explained why we need it, so I think we should keep it?

File: src/GraphQL/Utility/FileUpload.php
Comment (klausi) [inline src/GraphQL/Utility/FileUpload.php]: why do we need to change this code? We can just leave it like it was?

File: tests/modules/graphql_file_validate/graphql_file_validate.module
Comment (klausi) [inline tests/modules/graphql_file_validate/graphql_file_validate.module:15]: we need to keep this test and cannot remove it.

We need to rewrite this test to use FileValidationEvent as described in https://www.drupal.org/node/3363700

File: tests/src/Kernel/AlterableSchemaTest.php
Comment (klausi) [inline tests/src/Kernel/AlterableSchemaTest.php]: wrong conversion, this needs to be `->name()`. Then you also need to revert the changes to testEmptySchemaExtensionAlteredQueryResultPropertyAdded() above.

**Review by klausi — CHANGES_REQUESTED** [review]
Thanks a lot, looks very good already!

Added some comments that need to be fixed before we can merge this.

File: tests/src/Kernel/AlterableSchemaTest.php
Comment (vishalkhode1) [reply tests/src/Kernel/AlterableSchemaTest.php]: @klausi I think this was changed because of following reason:

It throws an error with Drupal 11, because the `getName()` method exists until [PHPUnit 9.5](https://github.com/sebastianbergmann/phpunit/blob/9.6/src/Framework/TestCase.php#L1035), but it doesn't exists on [PHPUnit 10](https://github.com/sebastianbergmann/phpunit/blob/10.5/src/Framework/TestCase.php).
So, as an alternative, we should use `$this->toString()` method.

File: .github/workflows/testing.yml
Comment (apathak18) [reply .github/workflows/testing.yml:11]: @klausi For some reason it failing the pipeline for 11.0.x ~~ see: https://github.com/drupal-graphql/graphql/actions/runs/9996316696/job/27630424860?pr=1407#step:11:1257

Comment (vishalkhode1) [reply .github/workflows/testing.yml:11]: @klausi No, we can't use `ubuntu-latest` if we need to run PHPUnit tests on Drupal 11. Because as of today `ubuntu-latest` resolves to ubuntu 22.04. See [this](https://github.com/actions/runner-images?tab=readme-ov-file#available-images).
The Drupal 11 requires minimum SQLite 3.45 and ubuntu 24.04 (Ubuntu Noble Numbat) only supports SQLite 3.45 and above. See this: 
https://launchpad.net/ubuntu/+source/sqlite3
Screenshot 2024-08-02 at 4 06 06 PM

Comment (vishalkhode1) [reply .github/workflows/testing.yml:11]: See failing PHPUnit on Drupal 11 when ubuntu-latest is used: https://github.com/drupal-graphql/graphql/actions/runs/9995716069/job/27628492492#step:11:590

File: tests/modules/graphql_file_validate/graphql_file_validate.module
Comment (apathak18) [reply tests/modules/graphql_file_validate/graphql_file_validate.module:15]: Done

File: src/GraphQL/Utility/FileUpload.php
Comment (apathak18) [reply src/GraphQL/Utility/FileUpload.php]: Just added another check and included array_merge to avoid any issues but know I reverted with the previous logic.

File: src/GraphQL/Utility/FileUpload.php
Comment (apathak18) [reply src/GraphQL/Utility/FileUpload.php]: Done

File: .github/workflows/testing.yml
Comment (apathak18) [reply .github/workflows/testing.yml:27]: Done

File: .github/workflows/testing.yml
Comment (klausi) [reply .github/workflows/testing.yml:11]: I see, please add a code comment above like "# We cannot use ubuntu-latest right now as it still points to 22.04 and we need a newer database driver."

File: tests/src/Kernel/AlterableSchemaTest.php
Comment (klausi) [reply tests/src/Kernel/AlterableSchemaTest.php]: `toString` is still the wrong method, as it returns something different.

I'm working on the getName fix in https://github.com/drupal-graphql/graphql/pull/1415

**Comment by klausi** [comment]
Merged #1415, please update your branch from 8.x-4.x, now you don't need any changes for AlterableSchemaTest.php.

File: .github/workflows/testing.yml
Comment (klausi) [inline .github/workflows/testing.yml]: quote at the end should be removed.

File: tests/modules/graphql_file_validate/graphql_file_validate.module
Comment (klausi) [inline tests/modules/graphql_file_validate/graphql_file_validate.module]: please remove the module file and move this code directly into the even subscriber.

**Comment by apathak18** [comment]
@klausi All feedbacks are addressed ~~ also rebased & resolved the conflicts, now this one is ready for a review.

**Review by klausi — CHANGES_REQUESTED** [review]
thanks, 2 minor points remaining.

I will also manually test the new test subscriber if it would fail when the file does not exist.

**Comment by klausi** [comment]
Please don't do force pushes, makes it harder for me to update local test branches. All the commits here don't matter, we will squash merge them aynway at the end.

**Comment by apathak18** [comment]
Ohh I'll note this from ahead.
For now I'm done with all the feedbacks, please check.

**Review by klausi — APPROVED** [review]
Excellent, all looking good now. Thank you for your patience!

**Comment by apathak18** [comment]
> Excellent, all looking good now. Thank you for your patience!
Thanks
Quote of the day: Patience is a key to success!!

**Comment by vipin-mittal-acquia** [comment]
Hello @klausi, When can we expect stable release of [GraphQL](https://www.drupal.org/project/graphql)? Thanks!

**Pull request merged**
